main.go

```diff
@@ -1,196 +1,864 @@
-	for _, mfd := range md.GetMessageFields() {
-		if mfd.GetTypeName() == ".google.protobuf.FieldMask" {
-			return s
-		}
+var importPathReplacer = strings.NewReplacer(
+	".", "_",
+	"/", "_",
+	"-", "_",
+)
+
+func eqCopyOp(dst, src string) string {
+	return fmt.Sprintf("%s = %s", dst, src)
+}
+
+func deepCopyOp(dst, src string) string {
+	return fmt.Sprintf("deepCopy(&%s, &%s)", dst, src)
+}
+
+func buildMethods(buf *strings.Builder, md *protokit.Descriptor, mdMap map[string]*protokit.Descriptor) (map[string]string, error) {
+	imports := map[string]string{}
+
+	paths, err := appendPaths(make([]string, 0, len(md.GetMessageFields())), "", md, mdMap, nil)
+	if err != nil {
+		// TODO: Return error here once https://github.com/TheThingsIndustries/protoc-gen-fieldmask/issues/5 is resolved.
+		log.Printf("Failed to traverse `%s`: %s, skipping...", md.GetFullName(), err)
+		return nil, nil
 	}
-	out := append(s, md)
-	for _, sub := range md.GetMessages() {
-		if strings.HasSuffix(sub.GetName(), "Entry") {
-			continue
-		}
-		out = p.appendDescriptor(out, sub)
+
+	fmt.Fprintf(buf, `
+var _%sFieldPaths = [...]string{%s}
+
+func (*%s) FieldMaskPaths() []string {`,
+		md.GetName(), `"`+strings.Join(paths, `", "`)+`"`,
+		md.GetName(),
+	)
+
+	sort.Strings(paths)
+	if len(paths) == 0 {
+		fmt.Fprintf(buf, `
+	return nil`,
+		)
+	} else {
+		fmt.Fprintf(buf, `
+	ret := make([]string, len(_%sFieldPaths))
+	copy(ret, _%sFieldPaths[:])
+	return ret`,
+			md.GetName(),
+			md.GetName(),
+		)
 	}
-	return out
+
+	fmt.Fprintf(buf, `
 }
 
-func (p *plugin) Generate(in *plugin_go.CodeGeneratorRequest) (*plugin_go.CodeGeneratorResponse, error) {
-	descriptors := protokit.ParseCodeGenRequest(in)
+func (dst *%s) SetFields(src *%s, paths ...string) {
+	for _, path := range paths {
+		switch path {`,
+		md.GetName(), md.GetName(),
+	)
 
-	messages := make(map[string]*protokit.Descriptor)
-	for _, fd := range descriptors {
-		for _, md := range fd.GetMessages() {
-			p.registerMessage(messages, md)
+	for _, p := range paths {
+		fmt.Fprintf(buf, `
+		case "%s":
+			var nilPath bool`,
+			p)
+
+		sp := strings.Split(p, ".")
+
+		srcPath := "src"
+		fm := md
+		for i := 0; i < len(sp)-1; i++ {
+			fd := fm.GetMessageField(sp[i])
+			if fd.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+				panic(errors.New("Fieldmask for repeated field generated"))
+			}
+
+			goType := fieldTypeName(fd)
+
+			goName := generator.CamelCase(fd.GetName())
+			if v, ok := fd.OptionExtensions["gogoproto.customname"].(*string); ok {
+				goName = *v
+			}
+			if v, ok := fd.OptionExtensions["gogoproto.embed"].(*bool); ok && *v {
+				goName = goType
+			}
+
+			if fd.OneofIndex != nil {
+				oneOfType := fmt.Sprintf("%s_%s", fm.GetName(), goName)
+				if fm.GetMessage(goName) != nil {
+					oneOfType = fmt.Sprintf("%s_", oneOfType)
+				}
+
+				srcPath = fmt.Sprintf("%s.Get%s()", srcPath, goName)
+
+				fmt.Fprintf(buf, `
+			nilPath = nilPath || %s == nil`,
```

Comment:
Could you please use brackets to make this and `L275` more readable?:
`nilPath = (nilPath || (%s == nil))`

Reply:
No. `nilPath = nilPath || %s == nil` is much more readable than `nilPath = (nilPath || (%s == nil))`, but if @johanstokking agrees, I'll change that.